test/tls12groupselection_test.c

```diff
@@ -0,0 +1,201 @@
+        if (current_test_vector->server_groups != NULL)
+            TEST_true_or_end(SSL_set1_groups_list(serverssl, current_test_vector->server_groups));
+        TEST_true_or_end(SSL_set_min_proto_version(clientssl, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_set_min_proto_version(serverssl, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_set_max_proto_version(clientssl, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_set_max_proto_version(serverssl, TLS1_2_VERSION));
+        if (current_test_vector->preference == SERVER_PREFERENCE)
+            SSL_set_options(serverssl, SSL_OP_CIPHER_SERVER_PREFERENCE);
+    }
+
+    /* Creating a test connection */
+    TEST_true_or_end(create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE));
+
+    /*
+     * Checking that the negotiated group matches our expectation
+     * and must be identical on server- and client-side obviously
+     */
+    negotiated_group_client = SSL_get_negotiated_group(clientssl);
+    negotiated_group_server = SSL_get_negotiated_group(serverssl);
+    if (!TEST_int_eq(negotiated_group_client, negotiated_group_server))
+        goto end;
+    if (!TEST_int_eq(negotiated_group_client, OBJ_sn2nid(current_test_vector->expected_group)))
+        goto end;
+
+    ok = 1;
+
+end:
+    SSL_free(serverssl);
+    SSL_free(clientssl);
+    SSL_CTX_free(server_ctx);
+    SSL_CTX_free(client_ctx);
+    return ok;
+}
+#endif /* !defined(OPENSSL_NO_EC) */
+
+static int tls12groupselection_test(void)
+{
+    int testresult = 1; /* Assume the test will succeed */
+#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_TLS1_2)
+    unsigned long i;
+
+    /*
+     * Call the code under test, once such that the ssl object is used,
+     * once such that the ctx is usedIf any of the tests fail (= return 0),
+     * the end result will be 0 thanks to multiplication
+     */
+    for (i = 0;
+         i < sizeof(tls12groupselection_tests) / sizeof(tls12groupselection_tests[0]);
+         i++) {
+        int res = 0;
+
+        fprintf(stdout, "==> Running TLSv1.2 test %lu\n", i);
+        res = run_12groupselection(&tls12groupselection_tests[i],
+                                   WORK_ON_SSL_OBJECT);
+        if (!res)
+            fprintf(stdout, "====> [ERROR] TLSv1.2 test %lu with WORK_ON_SSL_OBJECT failed\n", i);
+        testresult *= res;
+        res = run_12groupselection(&tls12groupselection_tests[i],
+                                   WORK_ON_CONTEXT);
+        if (!res)
+            fprintf(stdout, "====> [ERROR] TLSv1.2 test %lu with WORK_ON_CONTEXT failed\n", i);
+        testresult *= res;
+    }
+
+#endif /* !defined(OPENSSL_NO_EC) */
+
+    return testresult;
+}
+
+int setup_tests(void)
+{
+
+#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_TLS1_2)
+    if (!TEST_ptr(cert = test_get_argument(0))
+            || !TEST_ptr(privkey = test_get_argument(1)))
+        return 0;
+#endif /* !defined(OPENSSL_NO_EC) */
+
+    ADD_TEST(tls12groupselection_test);
```

Comment:
Why not add this to sslapitest where all the other similar tests are?

Reply:
The new drop will remove the tls12groupselection_test completely, I would prefer to keep the tls13groupselection_test separate from sslapitest.c reason being is that file is massive now and I personally prefer to not increase the size of this file.